rewriter/ChainedSig.cc

```diff
@@ -0,0 +1,184 @@
+        auto *block = ast::cast_tree<ast::Block>(send->block);
+        auto blockBody = ast::cast_tree<ast::Send>(block->body);
+
+        // If the blockBody is not a send, then we have a sequence of instructions inside the signature block
+        if (blockBody == nullptr) {
+            if (auto e = ctx.beginError(block->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Malformed signature: cannot have multiple instructions inside a signature block");
+            }
+
+            return tree;
+        }
+
+        return buildReplacement(ctx, send, blockBody);
+    }
+
+    bool firstReceiverIsSig(ast::Send *send) {
+        do {
+            if (send->fun == core::Names::sig()) {
+                this->previousSigSend = send;
+                return true;
+            }
+            send = ast::cast_tree<ast::Send>(send->recv);
+        } while (send);
+
+        return false;
+    }
+
+    void checkDuplicates(core::MutableContext ctx, ast::Send *sigBlock, core::NameRef fun) {
+        while (sigBlock) {
+            if (sigBlock->fun == fun) {
+                if (auto e = ctx.beginError(sigBlock->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                    e.setHeader("Duplicate invocation of `{}` in signature declaration", fun.toString(ctx));
+                }
+            }
+
+            sigBlock = ast::cast_tree<ast::Send>(sigBlock->recv);
+        }
+    }
+
+    bool invalidChainedStatement(core::MutableContext ctx, ast::Send *send) {
+        if (send->fun == core::Names::params() || send->fun == core::Names::returns() ||
+            send->fun == core::Names::void_() || send->fun == core::Names::bind() ||
+            send->fun == core::Names::checked() || send->fun == core::Names::onFailure() ||
+            send->fun == core::Names::typeParameters()) {
+            auto receiver = ast::cast_tree<ast::Send>(send->recv);
+
+            while (receiver && receiver->fun != core::Names::sig()) {
+                receiver = ast::cast_tree<ast::Send>(receiver->recv);
+            }
+
+            if (receiver) {
+                this->previousSigSend = receiver;
+
+                if (auto e = ctx.beginError(send->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                    e.setHeader("Cannot use `{}` outside of a sig block", send->fun.toString(ctx));
+                }
+            }
+
+            return true;
+        }
+
+        return false;
+    }
+
+    ast::ExpressionPtr buildReplacement(core::MutableContext ctx, ast::Send *send, ast::Send *blockBody) {
+        ast::Send::ARGS_store args;
+        auto treeCopy = send->deepCopy();
+        auto sendCopy = ast::cast_tree<ast::Send>(treeCopy);
+
+        // Go through each part of the chained sig and make sure it's not duplicated inside the block
+        // E.g.: `sig.abstract { abstract.void }`
+        while (sendCopy && sendCopy->fun != core::Names::sig()) {
+            checkDuplicates(ctx, blockBody, sendCopy->fun);
```

Comment:
ooc, do you think we can clearly separate the "checking improper use" from the "mutating the tree" phase?

mixing those two modes of operation has been one of the most common sources of bugs in sorbet over the years.

Reply:
ah, it looks like checking and mutating is just really intertwined below. maybe not worth it.

Reply:
My initial approach was exactly that, first checking all invalid usage and then trying to build a replacement, but we end up having a lot more loops, so I changed it.

Out of curiosity, what kinds of bugs have been caused by this pattern?

Reply:
Usually it works like this:

- someone mutates the tree (usually to `std::move(...)` something out of it) early in the method than needed
- for some reason, we decide to return early
- the logic to compute the replacement for the thing that was `move`'d didn't fire